File: src/ruby/tormenta20/models/magia.rb
# frozen_string_literal: true

require "active_record"
require "json"

module Tormenta20
  module Models
    # Model representing a Magia (Spell) in Tormenta20.
    #
    # Spells are categorized by type (arcane, divine, universal), circle (1-5),
    # and school (abjuration, evocation, etc.).
    #
    # @example Find all arcane spells
    #   Magia.arcanas
    #
    # @example Find 3rd circle evocation spells
    #   Magia.by_circle("3").by_school("evoc")
    #
    # @!attribute [rw] id
    #   @return [String] Unique identifier (e.g., "bola_de_fogo")
    # @!attribute [rw] name
    #   @return [String] Spell name (e.g., "Bola de Fogo")
    # @!attribute [rw] type
    #   @return [String] Spell type: "arcana", "divina", or "universal"
    # @!attribute [rw] circle
    #   @return [String] Spell circle (1-5)
    # @!attribute [rw] school
    #   @return [String] Magic school code (abjur, adiv, conv, encan, evoc, ilus, necro, trans)
    # @!attribute [rw] execution
    #   @return [String] Casting time
    # @!attribute [rw] range
    #   @return [String] Spell range
    # @!attribute [rw] duration
    #   @return [String] Spell duration
    # @!attribute [rw] description
    #   @return [String] Full spell description
    class Magia < Base
      self.table_name = "magias"
      self.inheritance_column = nil

      validates :id, presence: true, uniqueness: true
      validates :name, presence: true
      validates :type, presence: true, inclusion: { in: %w[arcana divina universal] }
      validates :circle, presence: true
      validates :school, presence: true, inclusion: { in: %w[abjur adiv conv encan evoc ilus necro trans] }
      validates :execution, presence: true
      validates :range, presence: true
      validates :duration, presence: true
      validates :description, presence: true

      # @!group Scopes

      # @!method arcanas
      #   Filter arcane spells only.
      #   @return [ActiveRecord::Relation<Magia>]
      scope :arcanas, -> { where(type: "arcana") }

      # @!method divinas
      #   Filter divine spells only.
      #   @return [ActiveRecord::Relation<Magia>]
      scope :divinas, -> { where(type: "divina") }

      # @!method universais
      #   Filter universal spells only.
      #   @return [ActiveRecord::Relation<Magia>]
      scope :universais, -> { where(type: "universal") }

      # @!method by_circle(circle)
      #   Filter spells by circle.
      #   @param circle [String, Integer] The spell circle (1-5)
      #   @return [ActiveRecord::Relation<Magia>]
      scope :by_circle, ->(circle) { where(circle: circle.to_s) }

      # @!method by_school(school)
      #   Filter spells by magic school.
      #   @param school [String] School code (abjur, adiv, conv, encan, evoc, ilus, necro, trans)
      #   @return [ActiveRecord::Relation<Magia>]
      scope :by_school, ->(school) { where(school: school) }

      # @!method by_type(type)
      #   Filter spells by type.
      #   @param type [String] Spell type (arcana, divina, universal)
      #   @return [ActiveRecord::Relation<Magia>]
      scope :by_type, ->(type) { where(type: type) }

      # @!endgroup

      class << self
        # Get all spells.
        #
        # @return [ActiveRecord::Relation<Magia>] All spells
        def todas
          all
        end

        # Get divine spells (alias for divinas scope).
        #
        # @return [ActiveRecord::Relation<Magia>] Divine spells
        def divinas_list
          divinas
        end

        # Get arcane spells (alias for arcanas scope).
        #
        # @return [ActiveRecord::Relation<Magia>] Arcane spells
        def arcanas_list
          arcanas
        end

        # Get universal spells (alias for universais scope).
        #
        # @return [ActiveRecord::Relation<Magia>] Universal spells
        def universais_list
          universais
        end

        # Find spells by circle (Portuguese alias).
        #
        # @param circle [String, Integer] The spell circle
        # @return [ActiveRecord::Relation<Magia>] Spells of the given circle
        def do_circulo(circle)
          by_circle(circle)
        end

        # Find spells by school (Portuguese alias).
        #
        # @param school [String] The magic school code
        # @return [ActiveRecord::Relation<Magia>] Spells of the given school
        def da_escola(school)
          by_school(school)
        end
      end

      # Get target information as a structured hash.
      #
      # @return [Hash] Target info with :amount, :up_to, and :type keys
      def target_info
        {
          amount: target_amount,
          up_to: target_up_to&.positive?,
          type: target_type
        }
      end

      # Get effect details as a structured hash.
      #
      # @return [Hash, nil] Effect details or nil if no effect info exists
      def effect_details_info
        return nil unless effect_shape || effect_dimention || effect_size

        {
          shape: effect_shape,
          dimention: effect_dimention,
          size: effect_size,
          other_details: effect_other_details
        }.compact
      end

      # Get resistance information as a structured hash.
      #
      # @return [Hash, nil] Resistance info or nil if no resistance exists
      def resistence_info
        return nil unless resistence_effect || resistence_skill

        {
          effect: resistence_effect,
          skill: resistence_skill
        }.compact
      end

      # Get extra costs as a structured hash.
      #
      # @return [Hash, nil] Extra costs or nil if no extra costs exist
      def extra_costs_info
        return nil unless extra_costs_material_component

        {
          material_component: extra_costs_material_component,
          material_component_cost: extra_costs_material_cost,
          pm_debuff: extra_costs_pm_debuff,
          pm_sacrifice: extra_costs_pm_sacrifice
        }.compact
      end

      # Convert spell to a hash representation.
      #
      # @return [Hash] Hash with all spell attributes
      def to_h
        {
          id: id,
          name: name,
          type: type,
          circle: circle,
          school: school,
          execution: execution,
          execution_details: execution_details,
          range: range,
          target: target_info,
          effect: effect,
          effect_details: effect_details_info,
          area_effect: area_effect,
          counterspell: counterspell,
          duration: duration,
          duration_details: duration_details,
          resistence: resistence_info,
          extra_costs: extra_costs_info,
          description: description,
          enhancements: enhancements,
          effects: effects
        }.compact
      end

      # @api private
      def method_missing(method_name, *args)
        return super unless respond_to_missing?(method_name)

        send(method_name.to_s.gsub("?", ""), *args)
      end

      # @api private
      def respond_to_missing?(method_name, include_private = false)
        attributes.key?(method_name.to_s) || super
      end
    end
  end
end

File: src/ts/models/magia.ts
import { BaseModel } from "./base.js"
import { Query } from "../query.js"
import { getBookReference, type BookReference } from "./concerns/book_referenceable.js"

const TABLE = "magias"

/** Tradition of the spell. */
export type MagiaType = "arcana" | "divina" | "universal"

/**
 * Magic school abbreviations used in Tormenta20.
 *
 * `abjur` · `adiv` · `conv` · `encan` · `evoc` · `ilus` · `necro` · `trans`
 */
export type MagiaSchool = "abjur" | "adiv" | "conv" | "encan" | "evoc" | "ilus" | "necro" | "trans"

/**
 * A Magia (spell) in Tormenta20.
 *
 * @example
 * ```ts
 * Magia.arcanas().where("circle = ?", "3").all()
 * Magia.divinas().all()
 * Magia.bySchool("necro").all()
 * Magia.find("bola_de_fogo")
 * ```
 */
export class Magia extends BaseModel {
  get type(): MagiaType { return this._row.type as MagiaType }
  /** Circle (level) of the spell, stored as a string (e.g. `"1"`, `"2"`, `"3"`). */
  get circle(): string { return this._row.circle as string }
  get school(): MagiaSchool { return this._row.school as MagiaSchool }
  /** Casting time (e.g. `"padrão"`, `"completa"`). */
  get execution(): string { return this._row.execution as string }
  get executionDetails(): string | null { return (this._row.execution_details as string) ?? null }
  get range(): string { return this._row.range as string }
  get duration(): string { return this._row.duration as string }
  get durationDetails(): string | null { return (this._row.duration_details as string) ?? null }
  /** ID of the counterspell, or `null`. */
  get counterspell(): string | null { return (this._row.counterspell as string) ?? null }
  get description(): string { return this._row.description as string }
  /** JSON array of enhancement objects (aprimoramentos). */
  get enhancements(): unknown[] { return BaseModel.json(this._row.enhancements, []) }
  /** JSON array of mechanical effect objects. */
  get effects(): unknown[] { return BaseModel.json(this._row.effects, []) }

  /**
   * Returns target information, or `null` if the spell has no specific target count/type.
   */
  targetInfo(): { amount: number | null; upTo: boolean; type: string | null } | null {
    const amount = this._row.target_amount as number | null
    const type = this._row.target_type as string | null
    if (!amount && !type) return null
    return { amount: amount ?? null, upTo: !!(this._row.target_up_to), type: type ?? null }
  }

  /**
   * Returns resistance information (saving throw), or `null` if the spell allows no save.
   */
  resistenceInfo(): { effect: string; skill: string } | null {
    const e = this._row.resistence_effect as string
    const s = this._row.resistence_skill as string
    if (!e && !s) return null
    return { effect: e, skill: s }
  }

  /**
   * Returns extra casting costs (material components, PM sacrifice, etc.),
   * or `null` if none exist.
   */
  extraCostsInfo(): Record<string, unknown> | null {
    const fields = ["extra_costs_material_component", "extra_costs_material_cost", "extra_costs_pm_debuff", "extra_costs_pm_sacrifice"]
    const result: Record<string, unknown> = {}
    for (const f of fields) {
      if (this._row[f]) result[f.replace("extra_costs_", "")] = this._row[f]
    }
    return Object.keys(result).length > 0 ? result : null
  }

  /** Returns the sourcebook page reference, or `null` if not indexed. */
  bookReference(): BookReference | null { return getBookReference(this.id) }

  toH() {
    return {
      id: this.id, name: this.name, type: this.type, circle: this.circle, school: this.school,
      execution: this.execution, execution_details: this.executionDetails, range: this.range,
      duration: this.duration, duration_details: this.durationDetails, counterspell: this.counterspell,
      description: this.description, target: this.targetInfo(), resistence: this.resistenceInfo(),
      extra_costs: this.extraCostsInfo(), enhancements: this.enhancements, effects: this.effects,
    }
  }

  static query(): Query<Magia> { return BaseModel.makeQuery(TABLE, Magia) }
  static all(): Magia[] { return Magia.query().all() }
  static find(id: string): Magia | null { return Magia.query().find(id) }
  static first(): Magia | null { return Magia.query().first() }
  static last(): Magia | null { return Magia.query().last() }
  static count(): number { return Magia.query().count() }

  static byType(type: MagiaType): Query<Magia> { return Magia.query().where("type = ?", type) }
  static byCircle(circle: string): Query<Magia> { return Magia.query().where("circle = ?", circle) }
  static bySchool(school: MagiaSchool): Query<Magia> { return Magia.query().where("school = ?", school) }

  static arcanas(): Query<Magia> { return Magia.byType("arcana") }
  static divinas(): Query<Magia> { return Magia.byType("divina") }
  static universais(): Query<Magia> { return Magia.byType("universal") }
  static doCirculo(circle: string): Query<Magia> { return Magia.byCircle(circle) }
  static daEscola(school: MagiaSchool): Query<Magia> { return Magia.bySchool(school) }
}
